src/core/models/inputs/utxo_input.c

```diff
@@ -1,108 +1,290 @@
-    printf("[%s:%d] inputs count must be < 127\n", __func__, __LINE__);
+  if (utxo_inputs_count(*inputs) >= UTXO_INPUT_MAX_COUNT) {
+    printf("[%s:%d] inputs count exceeded max count %d\n", __func__, __LINE__, UTXO_OUTPUT_MAX_COUNT);
     return -1;
   }
 
-  utxo_input_ht *elm = utxo_inputs_find_by_id(inputs, tx_id);
-  if (elm) {
-    printf("[%s:%d] transaction ID exists\n", __func__, __LINE__);
+  // Input Type must denote a UTXO Input.
+  if (type != 0) {
+    printf("[%s:%d] unknown input type\n", __func__, __LINE__);
     return -1;
   }
 
-  elm = malloc(sizeof(utxo_input_ht));
-  if (elm == NULL) {
-    printf("[%s:%d] OOM\n", __func__, __LINE__);
+  // Check if transaction id already exist in list
+  if(utxo_inputs_find_by_id(*inputs, id) != NULL) {
+    printf("[%s:%d] transaction id already present in input list\n", __func__, __LINE__);
     return -1;
   }
-  memcpy(elm->tx_id, tx_id, TRANSACTION_ID_BYTES);
-  memcpy(elm->keypair.pub, pub, ED_PUBLIC_KEY_BYTES);
-  memcpy(elm->keypair.priv, priv, ED_PRIVATE_KEY_BYTES);
-  elm->output_index = index;
-  HASH_ADD(hh, *inputs, tx_id, TRANSACTION_ID_BYTES, elm);
-  return 0;
+
+  // Check if index already exist in list
+  if(utxo_inputs_find_by_index(*inputs, index) != NULL) {
+    printf("[%s:%d] index already present in input list\n", __func__, __LINE__);
+    return -1;
+  }
+
+  utxo_inputs_list_t *next = malloc(sizeof(utxo_inputs_list_t));
+  if (next) {
+    next->input = malloc(sizeof(utxo_input_t));
+    if (next->input) {
+      // Currently only UTXO input is supported
+      next->input->input_type = type;
+      memcpy(next->input->tx_id, id, TRANSACTION_ID_BYTES);
+      next->input->output_index = index;
+      LL_APPEND(*inputs, next);
+      return 0;
+    }
+  }
+
+  if (next) {
+    if (next->input) {
+      free(next->input);
+    }
+    free(next);
+  }
+
+  return -1;
 }
 
-int utxo_inputs_add(utxo_input_ht **inputs, byte_t id[], uint16_t index) {
-  if (index > UTXO_INPUT_MAX_INDEX) {
+uint16_t utxo_inputs_count(utxo_inputs_list_t *inputs) {
+  utxo_inputs_list_t *elm = NULL;
+  uint16_t len = 0;
+  if (inputs) {
+    LL_COUNT(inputs, elm, len);
+  }
+  return len;
+}
+
+utxo_input_t *utxo_inputs_find_by_id(utxo_inputs_list_t *inputs, byte_t id[]) {
+  if (inputs == NULL) {
+    printf("[%s:%d] empty input list\n", __func__, __LINE__);
+    return NULL;
+  }
+
+  if(id == NULL) {
+    printf("[%s:%d] invalid transaction id\n", __func__, __LINE__);
+    return NULL;
+  }
+
+  uint16_t count = 0;
+  utxo_inputs_list_t *elm;
+  if (inputs) {
+    LL_FOREACH(inputs, elm) {
+      if (memcmp(elm->input->tx_id, id, TRANSACTION_ID_BYTES) == 0) {
+        return elm->input;
+      }
+    }
+  }
+  return NULL;
+}
+
+utxo_input_t *utxo_inputs_find_by_index(utxo_inputs_list_t *inputs, uint16_t index) {
+  if (inputs == NULL) {
+    printf("[%s:%d] empty input list\n", __func__, __LINE__);
+    return NULL;
+  }
+
+  if (index <= 0 || index >= UTXO_OUTPUT_MAX_COUNT) {
```

Comment:
Index can be 0 but can't be a negative number:
```suggestion
  if (index >= UTXO_OUTPUT_MAX_COUNT) {
```

Reply:
Done.